### PR TITLE
[ZingMp3] Improve extractor

### PR DESCRIPTION
**Pull request by hatienl0i261299**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Add new support for chart, user and update API to get streaming for video

    ZingMp3ChartHomeIE
    ZingMp3WeekChartIE
    ZingMp3ChartMusicVideoIE
    ZingMp3UserIE

### CODE REVIEW

File: yt_dlp/extractor/zingmp3.py [hunk 1]
```diff
@@ -31,7 +45,10 @@ def _extract_item(self, item, song_id, type_url, fatal):
         title = item.get('title') or item.get('alias')
 
         if type_url == 'video-clip':
+            info = self._download_json('http://api.mp3.zing.vn/api/mobile/video/getvideoinfo?requestdata={"id":"{%s}"}' % item_id, item_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:48]: use `query` and `json.dumps`

File: yt_dlp/extractor/zingmp3.py [hunk 2]
```diff
@@ -1,11 +1,14 @@
+import functools
 import hashlib
 import hmac
-import urllib.parse
 
 from .common import InfoExtractor
+from ..compat import compat_urllib_parse_urlencode
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:6]: urllib.parse.urlencode

File: yt_dlp/extractor/zingmp3.py [hunk 3]
```diff
@@ -212,11 +233,188 @@ class ZingMp3AlbumIE(ZingMp3BaseIE):
...
+            '_type': 'playlist',
+            'id': 'IWZ9Z0BO',
+            'title': 'zing-chart-korea',
+        },
+        'playlist_mincount': 10,
+    }]
+
+    def _process_data(self, data, chart_id, type_url):
+        items = data.get('items')
+        return self.playlist_result(self._entries(items), chart_id, f'zing-chart-{data.get("country", "")}')
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:306]:
```suggestion
        return self.playlist_result(
            self._entries(data['items']), chart_id, f'zing-chart-{data.get("country", "")}')
```

File: yt_dlp/extractor/zingmp3.py [hunk 4]
```diff
@@ -212,11 +233,188 @@ class ZingMp3AlbumIE(ZingMp3BaseIE):
...
+    def _fetch_page(self, song_id, type_url, page):
+        page += 1
+        api = self.get_api_with_signature(name_api=self._SLUG_API[type_url], param={
+            'id': song_id,
+            'type': 'genre',
+            'page': page,
+            'count': self._PER_PAGE
+        })
+        data = self._download_json(api, song_id)['data']
+        return self._entries(data.get('items'))
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:332]: why use `.get` when `_entries` will throw on None anyway?

File: yt_dlp/extractor/zingmp3.py [hunk 5]
```diff
@@ -212,11 +233,188 @@ class ZingMp3AlbumIE(ZingMp3BaseIE):
...
+                            (?:/
+                                (?P<type>bai-hat|single|album|video)
+                            )
+                            /?(?:[?#].*)?$
+                    '''
+    IE_NAME = 'zingmp3:user'
+    _TESTS = [{
+        'url': 'https://zingmp3.vn/Mr-Siro/bai-hat',
+        'info_dict': {
+            '_type': 'playlist',
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:354]:
```suggestion
```
same everywhere

File: yt_dlp/extractor/zingmp3.py [hunk 6]
```diff
@@ -212,11 +233,188 @@ class ZingMp3AlbumIE(ZingMp3BaseIE):
...
+        page += 1
+        name_api = self._SLUG_API['user-list-song'] if type_url == 'bai-hat' else self._SLUG_API['user-list-video']
+        api = self.get_api_with_signature(name_api=name_api, param={
+            'id': user_id,
+            'type': 'artist',
+            'page': page,
+            'count': self._PER_PAGE
+        })
+        data = self._download_json(api, user_id, query={'sort': 'new', 'sectionId': 'aSong'})['data']
+        return self._entries(data.get('items'))
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:399]: same as above

File: yt_dlp/extractor/zingmp3.py [hunk 7]
```diff
@@ -212,11 +233,188 @@ class ZingMp3AlbumIE(ZingMp3BaseIE):
...
+
+    def _real_extract(self, url):
+        user_alias, type_url = self._match_valid_url(url).group('user', 'type')
+        if not type_url:
+            type_url = 'bai-hat'
+        user_info = self._download_json(self.get_api_with_signature(name_api=self._SLUG_API['info-artist'], param={}),
+                                        video_id=user_alias, query={'alias': user_alias})['data']
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:406]:
```suggestion
        user_info = self._download_json(
            self.get_api_with_signature(name_api=self._SLUG_API['info-artist'], param={}),
            video_id=user_alias, query={'alias': user_alias})['data']
```

File: yt_dlp/extractor/zingmp3.py [hunk 8]
```diff
@@ -212,11 +233,188 @@ class ZingMp3AlbumIE(ZingMp3BaseIE):
...
+            entries = OnDemandPagedList(functools.partial(self._fetch_page, user_id, type_url), self._PER_PAGE)
+            return self.playlist_result(entries, user_id, f'{user_info.get("name")} - {type_url}', biography)
+        else:
+            lst_entries = []
+            for section in user_info.get('sections', {}):
+                if section.get('link') == f'/{user_alias}/{type_url}':
+                    items = section.get('items')
+                    for item in items:
+                        lst_entries.append(self.url_result(urljoin(self._DOMAIN, item.get('link'))))
+                    break
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:419]:
```suggestion
            entries = [
                self.url_result(urljoin(self._DOMAIN, link))
                for link in traverse_obj(user_info, (
                    'sections', lambda _, v: v['link'] == f'/{user_alias}/{type_url}', 'items', 'link'))]
```
untested

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/zingmp3.py [hunk 9]
```diff
@@ -212,11 +233,188 @@ class ZingMp3AlbumIE(ZingMp3BaseIE):
...
-                    yield entry
-
-        return self.playlist_result(entries(), traverse_obj(data, 'id', 'encodeId'),
+        items = traverse_obj(data, ('song', 'items')) or []
+        return self.playlist_result(self._entries(items), traverse_obj(data, 'id', 'encodeId'),
                                     traverse_obj(data, 'name', 'title'))
+
+
+class ZingMp3ChartHomeIE(ZingMp3BaseIE):
+    _VALID_URL = r'https?://(?:mp3\.zing|zingmp3)\.vn/(?P<id>(?:zing-chart|moi-phat-hanh))/?$'
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:242]:
```suggestion
    _VALID_URL = r'https?://(?:mp3\.zing|zingmp3)\.vn/(?P<id>(?:zing-chart|moi-phat-hanh))/?(?:[#?]|$)'
```

File: yt_dlp/extractor/zingmp3.py [hunk 10]
```diff
@@ -212,11 +233,188 @@ class ZingMp3AlbumIE(ZingMp3BaseIE):
...
+    def _process_data(self, data, chart_id, type_url):
+        if type_url == 'zing-chart':
+            items = traverse_obj(data, ('RTChart', 'items'), default=[])
+        else:
+            items = data.get('items')
+        return self.playlist_result(self._entries(items), type_url, type_url)
+
+
+class ZingMp3WeekChartIE(ZingMp3BaseIE):
+    _VALID_URL = r'https?://(?:mp3\.zing|zingmp3)\.vn/(?P<type>zing-chart-tuan)/.+?/(?P<id>[^\.]+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:276]: avoid `.` Same below

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/zingmp3.py (hunk 8 above)
Comment (hatienl0i261299) [reply yt_dlp/extractor/zingmp3.py:419]: No, its not work, and i think old code can be easier for reading

File: yt_dlp/extractor/zingmp3.py (hunk 10 above)
Comment (hatienl0i261299) [reply yt_dlp/extractor/zingmp3.py:276]: Sr, but could you explain more for it?

Comment (pukkandan) [reply yt_dlp/extractor/zingmp3.py:279]:
```suggestion
    _VALID_URL = r'https?://(?:mp3\.zing|zingmp3)\.vn/(?P<type>zing-chart-tuan)/[^/?#]+/(?P<id>\w+)'
```

File: yt_dlp/extractor/zingmp3.py [hunk 11]
```diff
@@ -120,6 +139,11 @@ def get_api_with_signature(self, name_api, param):
         }
         return f'{self._DOMAIN}{name_api}?{urllib.parse.urlencode(data)}'
 
+    def _entries(self, items):
+        for item in items:
+            if item:
+                yield self.url_result(urljoin(self._DOMAIN, item.get('link')))
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:145]:
```suggestion
            if item and item.get('link'):
                yield self.url_result(urljoin(self._DOMAIN, item['link']))
```
is this what u meant?

File: yt_dlp/extractor/zingmp3.py [hunk 12]
```diff
@@ -120,6 +139,11 @@ def get_api_with_signature(self, name_api, param):
         }
         return f'{self._DOMAIN}{name_api}?{urllib.parse.urlencode(data)}'
 
+    def _entries(self, items):
+        for item in items:
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:143]:
```suggestion
        for item in items or []:
```
see below comment

File: yt_dlp/extractor/zingmp3.py [hunk 13]
```diff
@@ -212,11 +236,189 @@ class ZingMp3AlbumIE(ZingMp3BaseIE):
...
+        api = self.get_api_with_signature(name_api=self._SLUG_API[type_url], param={
+            'id': song_id,
+            'type': 'genre',
+            'page': page,
+            'count': self._PER_PAGE
+        })
+        data = self._download_json(api, song_id)['data']
+        items = data.get('items')
+        if items:
+            return self._entries(items)
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:337]:
```suggestion
        return self._entries(data.get('items'))
```
see above comment

File: yt_dlp/extractor/zingmp3.py [hunk 14]
```diff
@@ -212,11 +236,189 @@ class ZingMp3AlbumIE(ZingMp3BaseIE):
...
+
+    def _real_extract(self, url):
+        song_id, type_url = self._match_valid_url(url).group('id', 'type')
+        entries = OnDemandPagedList(functools.partial(self._fetch_page, song_id, type_url), self._PER_PAGE)
+        return self.playlist_result(entries, song_id, 'zing-chart-video', 'zing-chart-video')
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:342]: as I asked in other PRs, why set a title that is useless?

File: yt_dlp/extractor/zingmp3.py [hunk 15]
```diff
@@ -212,11 +236,189 @@ class ZingMp3AlbumIE(ZingMp3BaseIE):
...
+
+class ZingMp3UserIE(ZingMp3BaseIE):
+    _VALID_URL = r'''(?x)
+                        https?://
+                            (?:mp3\.zing|zingmp3)\.vn/
+                            (?P<user>[^/]+)
+                            (?:/
+                                (?P<type>bai-hat|single|album|video)
+                            )
+                            /?(?:[?#].*)?$
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:353]:
```suggestion
                            /?(?:[?#]|$)
```

File: yt_dlp/extractor/zingmp3.py [hunk 16]
```diff
@@ -212,11 +236,189 @@ class ZingMp3AlbumIE(ZingMp3BaseIE):
...
+        return self.playlist_result(entries, song_id, 'zing-chart-video', 'zing-chart-video')
+
+
+class ZingMp3UserIE(ZingMp3BaseIE):
+    _VALID_URL = r'''(?x)
+                        https?://
+                            (?:mp3\.zing|zingmp3)\.vn/
+                            (?P<user>[^/]+)
+                            (?:/
+                                (?P<type>bai-hat|single|album|video)
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:351]:
```suggestion
                            (?:
                                /(?P<type>bai-hat|single|album|video)
```

File: yt_dlp/extractor/zingmp3.py [hunk 17]
```diff
@@ -212,11 +236,189 @@ class ZingMp3AlbumIE(ZingMp3BaseIE):
...
+        api = self.get_api_with_signature(name_api=name_api, param={
+            'id': user_id,
+            'type': 'artist',
+            'page': page,
+            'count': self._PER_PAGE
+        })
+        data = self._download_json(api, user_id, query={'sort': 'new', 'sectionId': 'aSong'})['data']
+        items = data.get('items')
+        if items:
+            return self._entries(items)
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:402]: similar to above

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/zingmp3.py (hunk 5 above)
Comment (pukkandan) [reply yt_dlp/extractor/zingmp3.py]: This is still left in some places

File: yt_dlp/extractor/zingmp3.py (hunk 14 above)
Comment (hatienl0i261299) [reply yt_dlp/extractor/zingmp3.py]: Fixed.

**Comment by hatienl0i261299** [comment]
@pukkandan Do i need update anything else, can we merge this PR?

**Comment by pukkandan** [comment]
The whole file is a mess and needs refactoring

**Pull request merged**